cli/spec.py

```diff
@@ -0,0 +1,244 @@
+        if arg_value is None and upper_arg_name in os.environ:
+            defaults[arg_name] = os.getenv(upper_arg_name)
+
+    # override defaults with the ini file args
+    if 'from-file' in res_args:
+        file_args = res_args['from-file']
+        if file_args is not None and res_args['command0'] in file_args:
+            defaults = utils.dict_merge(
+                file_args[res_args['command0']],
+                defaults,
+                conflict_resolver=utils.ConflictResolver.none_resolver)
+
+    # replace defaults with cli
+    utils.dict_merge(res_args, defaults,
+                     conflict_resolver=utils.ConflictResolver.none_resolver)
+
+    # check if we have all the required arguments defined at the end of merge
+    unset_args = []
+    only_args = []
+    for arg_name, arg_param in command_args.iteritems():
+        if 'required' in arg_param and \
+                arg_param['required'] and \
+                res_args.get(arg_name, None) is None:
+            unset_args.append(arg_name)
+        if 'requires_only' in arg_param:
+            only_args.extend(arg_param['requires_only'])
+    if len(set(unset_args).intersection(set(only_args))) > 0:
+        raise exceptions.IRConfigurationException(
+                "Required input arguments {} are not set!".format(unset_args))
+
+    return res_args
+
+# ***************************************************************
+# private methods
+# ***************************************************************
+
+
+def __load_spec_as_dict(module_name, config):
+    """
+    Loads the spec file as dict.
+
+    Returns tuple:
+        dict of options and groups for clg
+        dict of options without subcommands
+        dict of options for subcommands
+    """
+    clg_options = _get_specs(module_name, config)
+    global_options, subcommands_options = __process_options(clg_options)
+    return clg_options, global_options, subcommands_options
+
+
+def __process_options(spec_dict, lookup=False, trim=False):
+    """
+    Goes through all the spec options modifies them by removing
+     some options parameters (like defaults)
+    """
+    global_options = {}
+    subcommand_options = {}
+    __get_flat_options(global_options, spec_dict, lookup, trim)
+    __get_groups_options(global_options, spec_dict, lookup, trim)
+    # collect sub parser global_options
+    if 'subparsers' in spec_dict:
+        for subparser_name, params in spec_dict['subparsers'].iteritems():
+            subcommand_options[subparser_name], _ = __process_options(
+                params, lookup=True, trim=True)
+
+    return global_options, subcommand_options
+
+
+def __get_groups_options(options, spec_dict, lookup, trim):
+    """
+    Gets the dict of options nested within the spec groups
+    """
+    if 'groups' in spec_dict:
+        for group in spec_dict['groups']:
+            __get_flat_options(options, group, lookup, trim)
+
+
+def __get_flat_options(options, spec_dict, lookup, trim):
```

Comment:
Usually functions with name that starts with 'get' supposed to return some calculated value.